### PR TITLE
feat: implement hash wrappers

### PR DESCRIPTION
**Pull request by chokobole**
# Description

This PR implements [PaddingFreeSponge](https://github.com/Plonky3/Plonky3/blob/fde81db95b8eeb39e07890f4099a7d59daeec52f/symmetric/src/sponge.rs#L10-L44) and [TruncatedPermutation](https://github.com/Plonky3/Plonky3/blob/fde81db95b8eeb39e07890f4099a7d59daeec52f/symmetric/src/compression.rs#L6-L44). Both of them are needed to test [mmcs](https://github.com/Plonky3/Plonky3/blob/fde81db95b8eeb39e07890f4099a7d59daeec52f/merkle-tree/src/mmcs.rs#L188-L223). See https://github.com/Plonky3/Plonky3/blob/fde81db95b8eeb39e07890f4099a7d59daeec52f/merkle-tree/src/mmcs.rs#L183-L184

### CODE REVIEW

**Review by dongchangYoo — CHANGES_REQUESTED** [review]

**Comment by TomTaehoonKim** [comment]
@chokobole Could you fix commit message of 8f75095? From "a `uint32_t`" -> "`uint32_t`"?

File: tachyon/crypto/hashes/hasher.h [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+#ifndef TACHYON_CRYPTO_HASHES_HASHER_H_
+#define TACHYON_CRYPTO_HASHES_HASHER_H_
+
+namespace tachyon::crypto {
+
+template <typename Derived>
+class Hasher {
+ public:
+  // TODO(chokobole): Make this accept iterator as arguments.
```
Comment (ashjeong) [inline tachyon/crypto/hashes/hasher.h:14]:
```suggestion
  // TODO(chokobole): Make this accept iterator as an argument.

File: tachyon/crypto/hashes/sponge/padding_free_sponge.h [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+#ifndef TACHYON_CRYPTO_HASHES_SPONGE_PADDING_FREE_SPONGE_H_
+#define TACHYON_CRYPTO_HASHES_SPONGE_PADDING_FREE_SPONGE_H_
+
+#include <stddef.h>
+
+#include <array>
+#include <utility>
+
+#include "tachyon/base/logging.h"
```
Comment (ashjeong) [inline tachyon/crypto/hashes/sponge/padding_free_sponge.h:14]: Unneeded! Please remove this line & relevant line in the bezel file as well.

**Comment by TomTaehoonKim** [comment]
@chokobole Sorry but could you also add link to the reference for 63a2300?

File: tachyon/crypto/hashes/compressor.h [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+#ifndef TACHYON_CRYPTO_HASHES_COMPRESSOR_H_
+#define TACHYON_CRYPTO_HASHES_COMPRESSOR_H_
+
+namespace tachyon::crypto {
+
+template <typename Derived>
+class Compressor {
+ public:
+  // TODO(chokobole): Make this accept iterator as arguments.
```
Comment (ashjeong) [inline tachyon/crypto/hashes/compressor.h:14]:
```suggestion
  // TODO(chokobole): Make this accept iterator as an argument.

File: tachyon/crypto/hashes/sponge/truncated_permutation.h [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+#ifndef TACHYON_CRYPTO_HASHES_SPONGE_TRUNCATED_PERMUTATION_H_
+#define TACHYON_CRYPTO_HASHES_SPONGE_TRUNCATED_PERMUTATION_H_
+
+#include <stddef.h>
+
+#include <array>
+#include <utility>
+
+#include "tachyon/base/logging.h"
```
Comment (ashjeong) [inline tachyon/crypto/hashes/sponge/truncated_permutation.h:14]: Unneeded! Please remove this line & relevant line in the bazel file as well.

**Review by ashjeong — COMMENTED** [review]
It's quite interesting how `TruncatedPermutation` and `PaddingFreeSponge` are so similar

**Review by dongchangYoo — APPROVED** [review]
LGTM

File: tachyon/math/finite_fields/baby_bear/poseidon2.h [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        BabyBear{0x6f53160e}, BabyBear{0x273b11d1}, BabyBear{0x03176c5d},
+        BabyBear{0x72f9bbf9}, BabyBear{0x73ceba91}, BabyBear{0x5cdef81d},
+        BabyBear{0x01393285}, BabyBear{0x46daee06}, BabyBear{0x065d7ba6},
+        BabyBear{0x52d72d6f}, BabyBear{0x05dd05e0}, BabyBear{0x3bab4b63},
+        BabyBear{0x6ada3842}, BabyBear{0x2fc5fbec}, BabyBear{0x770d61b0},
+        BabyBear{0x5715aae9}, BabyBear{0x03ef0e90}, BabyBear{0x75b6c770},
+        BabyBear{0x242adf5f}, BabyBear{0x00d0ca4c}, BabyBear{0x36c0e388},
+    };
+  } else {
+    static_assert(base::AlwaysFalse<std::array<BabyBear, N>>);
```
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/math/finite_fields/baby_bear/poseidon2.h:40]: Actually, this should be without `()`.

File: tachyon/crypto/hashes/sponge/poseidon2/poseidon2_unittest.cc [hunk 6]
```diff
@@ -64,4 +68,30 @@ TEST_F(Poseidon2Test, Copyable) {
   EXPECT_EQ(value, expected);
 }
 
+namespace {
+
+class Poseidon2BabyBearTest : public math::FiniteFieldTest<math::BabyBear> {};
```
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/crypto/hashes/sponge/poseidon2/poseidon2_unittest.cc:72]: Yes, throughout the code we haven't actually tested for all the possible prime fields. We've chose mostly `bn254::Fr` for a EC-based prime field.

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

File: tachyon/crypto/hashes/sponge/poseidon2/poseidon2_horizen_internal_matrix.h [hunk 7]
```diff
@@ -3,8 +3,8 @@
 // can be found in the LICENSE-MIT.plonky3 and the LICENCE-APACHE.plonky3
 // file.
```
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/crypto/hashes/sponge/poseidon2/poseidon2_horizen_internal_matrix.h:4]: Actually, the way of computing is taken from Plonky3. Please refer to [horizen](https://github.com/HorizenLabs/poseidon2/blob/bb476b9ca38198cf5092487283c8b8c5d4317c4e/plain_implementations/src/poseidon2/poseidon2.rs#L165-L208) and [plonky3](https://github.com/Plonky3/Plonky3/blob/fde81db95b8eeb39e07890f4099a7d59daeec52f/poseidon2/src/diffusion.rs#L35-L45).

**Review by ashjeong — APPROVED** [review]
LGTM

**Review by Insun35 — APPROVED** [review]
LGTM

**Pull request merged**